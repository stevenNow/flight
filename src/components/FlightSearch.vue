<template>
  <div class="center">
    <a-row style="margin-top: 200px">
      <a-form-model :model="searchForm">
        <a-form-model-item label="Origin">
          <a-select v-model="searchForm.origin">
            <a-select-option value="LAX">
              Los Angeles
            </a-select-option>
            <a-select-option value="ORD">
              O'hare
            </a-select-option>
            <a-select-option value="JFK">
              John F Kennedy
            </a-select-option>
          </a-select>
        </a-form-model-item>
        <a-form-model-item label="Departure">
          <a-date-picker v-model="searchForm.departure" :disabled-date="disabledDate" />
        </a-form-model-item>
        <a-form-model-item label="Destination">
          <a-select  v-model="searchForm.destination">
            <a-select-option value="LAX">
              Los Angeles
            </a-select-option>
            <a-select-option value="ORD">
              O'hare
            </a-select-option>
            <a-select-option value="JFK">
              John F Kennedy
            </a-select-option>
          </a-select>
        </a-form-model-item>
        <a-form-model-item label="Return">
          <a-date-picker v-model="searchForm.return" :disabled-date="disabledDate" />
        </a-form-model-item>
        <a-form-model-item label="Cabin Type">
          <a-select v-model="searchForm.cabinType">
            <a-select-option value="economy">
              Economy
            </a-select-option>
            <a-select-option value="business">
              Business
            </a-select-option>
            <a-select-option value="firstClass">
              First Class
            </a-select-option>
          </a-select>
        </a-form-model-item>
        <a-form-model-item  :wrapper-col="{ span: 24 }">
          <a-button type="primary" @click="onSearch" :disabled="disabledForm">
            Search
          </a-button>
          <a-button  @click="resetForm" style="margin-left:10px;">
            Reset
          </a-button>
          </a-form-model-item>
      </a-form-model>
    </a-row>
  </div>
</template>

<script>
import {mapMutations, mapGetters} from "vuex";
import moment from "moment"
export default {
  name: "FlightSearch",
  props: {},
  components: {
  },
  computed: {
    disabledForm() {
      return !(
          this.searchForm.origin && this.searchForm.departure
          && this.searchForm.destination && this.searchForm.return && this.searchForm.cabinType
      )
    },
  },
  data() {
    return {
      searchForm: {
        requestId: "1234",
        origin: null,
        departure: null,
        destination: null,
        return: null,
        cabinType: null
      },
    };
  },
  created() {
    this.searchForm.origin = this.getOrigin();
    this.searchForm.departure = this.getDeparture();
    this.searchForm.destination = this.getDestination();
    this.searchForm.return = this.getReturn();
    this.searchForm.cabinType = this.getCabinType();
  },
  methods: {
    moment,
    ...mapMutations([
        "setRequestId", "setOrigin", "setDeparture", "setDestination", "setReturn", "setCabinType"
    ]),
    ...mapGetters([
      "getRequestId", "getOrigin", "getDeparture", "getDestination", "getReturn", "getCabinType"
    ]),
    onSearch(val) {
        this.searchTerm = val;
        //requirements ask to emit this event, but i don't use it. using vuex instead.
        this.$emit("flight-search-request", this.searchForm);
        this.setSearchValuesInStore();
        this.$router.push({path: "/flightResult"});
    },
    setSearchValuesInStore() {
      this.setRequestId(this.searchForm.requestId);
      this.setOrigin(this.searchForm.origin);
      this.setDeparture(this.searchForm.departure);
      this.setDestination(this.searchForm.destination);
      this.setReturn(this.searchForm.return);
      this.setCabinType(this.searchForm.cabinType);
    },
    disabledDate(current) {
      // Can not select days before today and today
      return current && current < moment().startOf('day') -1;
    },
    resetForm() {
      this.searchForm.origin = null;
      this.searchForm.departure = null;
      this.searchForm.destination = null;
      this.searchForm.return = null;
      this.searchForm.cabinType = null;
    }
  },
};
</script>

<style>
.center {
  margin: auto;
  width: 50%;
}
</style>
