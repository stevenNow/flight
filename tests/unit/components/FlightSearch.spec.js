import Vuex from "vuex";
import { createLocalVue, mount } from "@vue/test-utils";
import Antd from "ant-design-vue";
import FlightSearch from "../../../src/components/FlightSearch";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Antd);

describe("FlightSearch Test", () => {
  let getters;
  let mutations;
  let store;

  beforeEach(() => {
    getters = {
      getOrigin: jest.fn(),
      getDeparture: jest.fn(),
      getDestination: jest.fn(),
      getReturn: jest.fn(),
      getCabinType: jest.fn()
    };

    mutations = {};

    store = new Vuex.Store({
      getters,
      modules: {},
      state: {},
      mutations,
    });
  });

  it("Tests FlightSearch Functions", async () => {
    const wrapper = mount(FlightSearch, {
      store,
      localVue,
      mocks: {
        $router: {
          push: jest.fn(),
        },
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
    //test disabledSubmit
    expect(FlightSearch.computed.disabledSubmit.call(wrapper.vm)).toBeTruthy();
    wrapper.vm.searchForm.origin = "LAX";
    expect(FlightSearch.computed.disabledSubmit.call(wrapper.vm)).toBeTruthy();
    //test filteredDestinations
    expect(FlightSearch.computed.filteredDestinations.call(wrapper.vm))
        .toStrictEqual([{key: "JFK", label: "John F Kennedy"},
      {key: "ORD", label: "O'hare"}]);
    wrapper.vm.searchForm.departure = "1/1/2022";
    wrapper.vm.searchForm.destination = "ORD";
    //test filteredOrigins
    expect(FlightSearch.computed.filteredOrigins.call(wrapper.vm))
        .toStrictEqual([{key: "JFK", label: "John F Kennedy"},{key: "LAX", label: "Los Angeles"}
          ]);
    wrapper.vm.searchForm.return = "1/2/2022";
    wrapper.vm.searchForm.cabinType = "fancy";
    expect(FlightSearch.computed.disabledSubmit.call(wrapper.vm)).toBeFalsy();
    //test reset
    wrapper.vm.resetForm();
    expect(wrapper.vm.searchForm.origin).toBeNull();
    expect(wrapper.vm.searchForm.departure).toBeNull();
    expect(wrapper.vm.searchForm.destination).toBeNull();
    expect(wrapper.vm.searchForm.return).toBeNull();
    expect(wrapper.vm.searchForm.cabinType).toBeNull();
  });
});