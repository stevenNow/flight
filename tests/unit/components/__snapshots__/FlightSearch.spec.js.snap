// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightSearch Test Tests FlightSearch Functions 1`] = `
<div
  class="center"
>
  <div
    class="ant-row"
    style="margin-top: 200px;"
  >
    <form
      class="ant-form ant-form-horizontal"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            title="Origin"
          >
            Origin
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <div
                class="ant-select ant-select-enabled ant-select-allow-clear"
                tabindex="0"
              >
                <div
                  aria-autocomplete="list"
                  aria-controls="test-uuid"
                  aria-haspopup="true"
                  class="ant-select-selection ant-select-selection--single"
                  role="combobox"
                >
                  <div
                    class="ant-select-selection__rendered"
                  />
                  <span
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <i
                      aria-label="icon: down"
                      class="anticon anticon-down ant-select-arrow-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </i>
                  </span>
                </div>
              </div>
            </span>
            <transition-stub
              appear="true"
            />
          </div>
        </div>
      </div>
       
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            title="Departure"
          >
            Departure
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <span
                class="ant-calendar-picker"
                style="width: 100%;"
              >
                <div
                  class=""
                >
                  <input
                    class="ant-calendar-picker-input ant-input"
                    placeholder="Select date"
                    readonly="true"
                  />
                  <i
                    aria-label="icon: calendar"
                    class="anticon anticon-calendar ant-calendar-picker-icon"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="calendar"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                      />
                    </svg>
                  </i>
                </div>
              </span>
            </span>
            <transition-stub
              appear="true"
            />
          </div>
        </div>
      </div>
       
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            title="Destination"
          >
            Destination
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <div
                class="ant-select ant-select-enabled ant-select-allow-clear"
                tabindex="0"
              >
                <div
                  aria-autocomplete="list"
                  aria-controls="test-uuid"
                  aria-haspopup="true"
                  class="ant-select-selection ant-select-selection--single"
                  role="combobox"
                >
                  <div
                    class="ant-select-selection__rendered"
                  />
                  <span
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <i
                      aria-label="icon: down"
                      class="anticon anticon-down ant-select-arrow-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </i>
                  </span>
                </div>
              </div>
            </span>
            <transition-stub
              appear="true"
            />
          </div>
        </div>
      </div>
       
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            title="Return"
          >
            Return
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <span
                class="ant-calendar-picker"
                style="width: 100%;"
              >
                <div
                  class=""
                >
                  <input
                    class="ant-calendar-picker-input ant-input"
                    placeholder="Select date"
                    readonly="true"
                  />
                  <i
                    aria-label="icon: calendar"
                    class="anticon anticon-calendar ant-calendar-picker-icon"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="calendar"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                      />
                    </svg>
                  </i>
                </div>
              </span>
            </span>
            <transition-stub
              appear="true"
            />
          </div>
        </div>
      </div>
       
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            title="Cabin Type"
          >
            Cabin Type
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <div
                class="ant-select ant-select-enabled ant-select-allow-clear"
                tabindex="0"
              >
                <div
                  aria-autocomplete="list"
                  aria-controls="test-uuid"
                  aria-haspopup="true"
                  class="ant-select-selection ant-select-selection--single"
                  role="combobox"
                >
                  <div
                    class="ant-select-selection__rendered"
                  />
                  <span
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <i
                      aria-label="icon: down"
                      class="anticon anticon-down ant-select-arrow-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </i>
                  </span>
                </div>
              </div>
            </span>
            <transition-stub
              appear="true"
            />
          </div>
        </div>
      </div>
       
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-col-24 ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <button
                class="ant-btn ant-btn-primary"
                disabled="disabled"
                type="button"
              >
                <span>
                  Search
                </span>
              </button>
              <button
                class="ant-btn"
                style="margin-left: 10px;"
                type="button"
              >
                <span>
                  Reset
                </span>
              </button>
            </span>
            <transition-stub
              appear="true"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
